### PR TITLE
#168781714 add chat functionality

### PR DESCRIPTION
**Pull request by alainmateso**
#### What does this PR do?
Add chat functionality

#### Description of Task to be completed?
- Adds chat functionality for logged in users
- Adds the chat interface

#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Browse to the repo directory
- navigate to ft-users-chat-168781714 branch
- Run `npm install`
- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`

To chat

- In Postman, login and copy the token
- Paste the token into the local storage of the browser
- In the browser, navigate to localhost:3000
You should see an interface that provides the text input and send button
Type your messages, hit send, and your messages as well as those from everyone connected should appear in the console

To view the chat history

- In Postman Provide a login bearer token
- send a `GET` request to `http://localhost:<:port>/api/v1/chat`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#168781714
#### Screenshots (if appropriate)
Screen Shot 2019-11-06 at 18 06 15
Screen Shot 2019-11-06 at 18 06 33
Screen Shot 2019-11-06 at 18 07 35

#### Questions:
N/A

### CODE REVIEW

**Review by Cheza-Dzabala — APPROVED** [review]
The only concern is the coverage... But that aside, I approve.

**Review by sabin18 — CHANGES_REQUESTED** [review]
@alainmateso good job however It would be batter if you use username in place of email .

**Comment by alainmateso** [comment]
> @alainmateso good job however It would be batter if you use username in place of email .

Well noted and implemented. Thanks for the feedback

**Review by sabin18 — APPROVED** [review]
@alainmateso good job.

File: src/controllers/chatController.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/* eslint-disable */
+import jwt from 'jsonwebtoken';
+import models from '../database/models';
+import responseUtil from '../utils/responseUtil';
+import strings from '../utils/stringsUtil';
+
+const { EMPTY_HISTORY, CHAT_HISTORY } = strings.chat;
+
+const verifyToken = token => {
```
Comment (nakiwuge) [inline src/controllers/chatController.js]: rename this to `verifyToken` to `getLoggedInUser`

File: src/controllers/chatController.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      socket.emit('connection_error', error.message);
+    }
+  });
+
+  socket.on('chatMessage', async (chatObject) => {
+    const { user } = socket;
+    const { message } = chatObject;
+    const userInfo = await models.users.findOne({ where: { id: user.payload.id } })
+
+    io.emit('chatMessage', { sender: userInfo.username, message });
```
Comment (nakiwuge) [inline src/controllers/chatController.js]: Putt this in a try and catch

File: src/controllers/chatController.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  try {
+    const user = jwt.verify(token, process.env.JWT_SECRET);
+    return user;
+  } catch (error) {
+    throw new Error('Please login again');
+  }
+};
+
+const chat = (io, socket) => {
+  io.use((socket, next) => {
```
Comment (nakiwuge) [inline src/controllers/chatController.js]: Putt this in a try and catch

File: src/controllers/chatController.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+      const user = verifyToken(token);
+      socket.user = user;
+      next();
+    } catch (error) {
+      socket.emit('connection_error', error.message);
+    }
+  });
+
+  socket.on('chatMessage', async (chatObject) => {
```
Comment (nakiwuge) [inline src/controllers/chatController.js:33]: Putt this in a try and catch

File: src/index.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  <body>
+    <ul id="messages"></ul>
+    <form action="">
+      <input id="message" autocomplete="off" /><button>Send</button>
+    </form>
+    <script src="/socket.io/socket.io.js"></script>
+    <script src="https://code.jquery.com/jquery-1.11.1.js"></script>
+    <script>
+      $(() => {
+        const socket = io("http://localhost:3000", {
```
Comment (nakiwuge) [inline src/index.html]: Make the url dynamic

File: src/routes/api/index.js [hunk 6]
```diff
@@ -21,5 +22,6 @@ router.use('/ratings', ratingRoutes);
 router.use('/comments', comments);
 router.use('/api-docs', swaggerRoute);
 router.use('/destinations', destinationRoutes);
+router.use('/chat', chatRoutes);
```
Comment (nakiwuge) [inline src/routes/api/index.js]: rename this `/chat` to `/chats`

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@alainmateso great job on the implementation. Kindly check out the comments I have left on this pr

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**